driver-core/src/main/com/mongodb/operation/ChangeStreamBatchCursor.java

```diff
@@ -0,0 +1,146 @@
+        this.resumeToken = changeStreamOperation.getResumeToken();
+        this.wrapped = wrapped;
+        this.binding = binding.retain();
+    }
+
+    BatchCursor<RawBsonDocument> getWrapped() {
+        return wrapped;
+    }
+
+    @Override
+    public boolean hasNext() {
+        return resumeableOperation(new Function<BatchCursor<RawBsonDocument>, Boolean>() {
+            @Override
+            public Boolean apply(final BatchCursor<RawBsonDocument> queryBatchCursor) {
+                return queryBatchCursor.hasNext();
+            }
+        });
+    }
+
+    @Override
+    public List<T> next() {
+        return resumeableOperation(new Function<BatchCursor<RawBsonDocument>, List<T>>() {
+            @Override
+            public List<T> apply(final BatchCursor<RawBsonDocument> queryBatchCursor) {
+                return convertResults(queryBatchCursor.next());
+            }
+        });
+    }
+
+    @Override
+    public List<T> tryNext() {
+        return resumeableOperation(new Function<BatchCursor<RawBsonDocument>, List<T>>() {
+            @Override
+            public List<T> apply(final BatchCursor<RawBsonDocument> queryBatchCursor) {
+                return convertResults(queryBatchCursor.tryNext());
+            }
+        });
+    }
+
+    @Override
+    public void close() {
+        wrapped.close();
+        binding.release();
+    }
+
+    @Override
+    public void setBatchSize(final int batchSize) {
+        wrapped.setBatchSize(batchSize);
+    }
+
+    @Override
+    public int getBatchSize() {
+        return wrapped.getBatchSize();
+    }
+
+    @Override
+    public ServerCursor getServerCursor() {
+        return wrapped.getServerCursor();
+    }
+
+    @Override
+    public ServerAddress getServerAddress() {
+        return wrapped.getServerAddress();
+    }
+
+    @Override
+    public void remove() {
+        throw new UnsupportedOperationException("Not implemented yet!");
+    }
+
+    private List<T> convertResults(final List<RawBsonDocument> rawDocuments) {
+        List<T> results = null;
+        if (rawDocuments != null) {
+            results = new ArrayList<T>();
+            for (RawBsonDocument rawDocument : rawDocuments) {
+                if (!rawDocument.containsKey("_id")) {
+                    throw new MongoChangeStreamException("Cannot provide resume functionality when the resume token is missing.");
+                }
+                resumeToken = rawDocument.getDocument("_id");
+                results.add(rawDocument.decode(changeStreamOperation.getDecoder()));
+            }
+        }
+        return results;
+    }
+
+    <R> R resumeableOperation(final Function<BatchCursor<RawBsonDocument>, R> function) {
+        try {
+            return function.apply(wrapped);
+        } catch (MongoNotPrimaryException e) {
+            // Ignore
+        } catch (MongoCursorNotFoundException w) {
+            // Ignore
+        } catch (MongoSocketReadException e) {
```

Comment:
This should also be MongoSocketException. Otherwise we exclude socket open and socket write exceptions.